### PR TITLE
Update lincheck to 2.5

### PR DESCRIPTION
**Pull request by ndkoval**
Update lincheck to 2.5 and re-write tests with `suspend` operations using it

### CODE REVIEW

**Comment by ndkoval** [comment]
It also fixes a linearizability bug in `ArrayChannel.isClosedForReceive`: #1530

File: kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    Param(name = "value", gen = IntGen::class, conf = "1:5"),
+    Param(name = "closeToken", gen = IntGen::class, conf = "1:3")
+)
+class ChannelsLCStressTestImpl: VerifierState() {
+    private val c = channelKind.create()
+
+    @Operation
+    suspend fun send(@Param(name = "value") value: Int) = try {
+        c.send(value)
+    } catch (e : NumberedCancellationException) {
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt]: Style nit: No space before `:` when used to specify variable type (in all the other cases in this file, too)

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt [hunk 2]
```diff
@@ -5,89 +5,54 @@
...
+            .check(this::class)
     }
-
-    override fun hashCode(): Int = Snapshot(q).hashCode()
 }
 
-@Param(name = "value", gen = IntGen::class, conf = "1:3")
-class MCLockFreeTaskQueueLCStressTest : LockFreeTaskQueueLCTestBase() {
-    private val q: LockFreeTaskQueue<Int> = LockFreeTaskQueue(singleConsumer = false)
+private var singleConsumer = false
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt]: I don't see any need to make it top-level var add `singleConsumer = when (this::class) { ... }` above. It can be a constructor parameter to `LockFreeTaskQueueWithoutRemoveLCStressTest` class that both inherited classes (`MCLockFreeTaskQueueWithRemoveLCStressTest` and `SCLockFreeTaskQueueWithRemoveLCStressTest`) that are actually being tested pass an appropriate value of.

File: kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt [hunk 3]
```diff
@@ -40,7 +40,7 @@ internal open class ArrayChannel<E>(
     private var size: Int = 0 // Invariant: size <= capacity
 
     protected final override val isBufferAlwaysEmpty: Boolean get() = false
-    protected final override val isBufferEmpty: Boolean get() = size == 0
+    protected final override val isBufferEmpty: Boolean get() = lock.withLock { size == 0 }
```
[the remark this answers is not in the document]
Comment (ndkoval) [reply kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt]: I would not be so scared since
1. this is actually not a _fast path_ because it suspends and suspensions are expensive
2. lock-based channel implementations in Golang are not worse in practice
3. this code will be completely removed with the new fast and scalable channels
4. this extra lock acquisition does not affect uncontended scenarios 

At the same time, your suggestion breaks the semantics of `isBufferEmpty` and makes the code harder to understand.

File: kotlinx-coroutines-core/common/test/TestBase.common.kt [hunk 4]
```diff
@@ -10,10 +10,10 @@ import kotlinx.coroutines.flow.*
 import kotlin.coroutines.*
 import kotlin.test.*
 
-public expect open class TestBase constructor() {
-    public val isStressTest: Boolean
-    public val stressTestMultiplier: Int
+public expect val isStressTest: Boolean
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/test/TestBase.common.kt]: Please, extract this into a separate commit and a separate PR. It would great to have it asap and separately.

File: kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ndkoval) [reply kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt]: >  Moreover, it looks scary, since this is an atomic multi-word operation that other threads might be helping with.
This change adds only one read under the lock -- seems like a safe change

File: kotlinx-coroutines-core/common/test/TestBase.common.kt (hunk 4 above)
Comment (ndkoval) [reply kotlinx-coroutines-core/common/test/TestBase.common.kt]: See #1579

File: kotlinx-coroutines-core/jvm/test/LCStressOptionsDefault.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+package kotlinx.coroutines
+
+import org.jetbrains.kotlinx.lincheck.*
+import org.jetbrains.kotlinx.lincheck.strategy.stress.*
+import kotlin.reflect.*
+
+class LCStressOptionsDefault : StressOptions() {
+    init {
+        iterations(100 * stressTestMultiplierSqrt)
+        invocationsPerIteration(20_000 * stressTestMultiplierSqrt)
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/LCStressOptionsDefault.kt]: This makes tests too slow on regular builds. See https://teamcity.jetbrains.com/viewLog.html?buildId=2571931&buildTypeId=KotlinTools_KotlinxCoroutines_BuildAggregated&tab=testsInfo
For example, `LockFreeListLCStressTest.testAddRemoveLinearizability` now takes 23 seconds.

File: kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Operation
+    fun isClosedForReceive() = c.isClosedForReceive
+
+    @Operation
+    fun isClosedForSend() = c.isClosedForSend
+
+    // TODO: this operation should be (and can be!) linearizable, but is not
+    // @Operation
+    fun isEmpty() = c.isEmpty
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt]: Can we still test `isEmpty` for rendezvous channels where it should be linearizable? It is not a short-stopper if we cannot, but would be a nice thing to have.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt [hunk 7]
```diff
@@ -5,89 +5,48 @@
...
 
     @Operation
     fun addLast(@Param(name = "value") value: Int) = q.addLast(value)
 
-    /**
-     * Note that removeFirstOrNull is not linearizable w.r.t. to addLast, so here
-     * we test only linearizability of close.
-     */
-//    @Operation(group = "consumer")
-//    fun removeFirstOrNull() = q.removeFirstOrNull()
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt]: Please, leave this commented out section together with comment in the code.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt [hunk 8]
```diff
@@ -5,89 +5,48 @@
...
+    fun testWithoutRemove() = LCStressOptionsDefault().check(this::class)
 
-        other as SCLockFreeTaskQueueLCStressTest
-
-        return Snapshot(q) == Snapshot(other.q)
+    @Test
+    fun testWithRemoveForQuiescentConsistency() {
+        LCStressOptionsDefault()
+            .verifier(QuiescentConsistencyVerifier::class.java)
+            .actorsPerThread(if (isStressTest) 3 else 2)
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt]: This seems to be a default `actorsPerThread` setting and it can be safely removed from here.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt [hunk 9]
```diff
@@ -5,16 +5,16 @@
...
 import kotlinx.coroutines.internal.*
+import org.jetbrains.kotlinx.lincheck.annotations.*
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.paramgen.*
+import org.jetbrains.kotlinx.lincheck.verifier.*
 import kotlin.test.*
 
-@Param(name = "value", gen = IntGen::class, conf = "1:3")
-class LockFreeListLCStressTest : TestBase() {
+@Param(name = "value", gen = IntGen::class, conf = "1:5")
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: This test should use a narrow range of values (`1:3`), because the key operation whose linearizability is checked here is `removeFirstOrPeekIfNotSame` which must have a high change of using the same value as the recently added nodes.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt [hunk 10]
```diff
@@ -44,29 +44,11 @@ class LockFreeListLCStressTest : TestBase() {
...
-        return curElements == other.curElements
-    }
+    fun testAddRemoveLinearizability() = LCStressOptionsDefault().check(this::class)
 
-    override fun hashCode(): Int {
-        return curElements.hashCode()
+    override fun extractState(): Any {
+        val elements = ArrayList<Int>()
+        q.forEach<Node> { elements.add(it.value) }
+        return elements
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: It would be also very useful to add call to `q.validate()` here which also checks that the resulting doubly-linked list has consistent links in both directions.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt [hunk 11]
```diff
@@ -44,29 +44,11 @@ class LockFreeListLCStressTest : TestBase() {
...
-            q.forEach<Node> { _curElements!!.add(it.value) }
-        }
-        return _curElements!!
-    }
-
-    override fun equals(other: Any?): Boolean {
-        other as LockFreeListLCStressTest
-        return curElements == other.curElements
-    }
+    fun testAddRemoveLinearizability() = LCStressOptionsDefault().check(this::class)
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: Note, that previous version of LinCheck was running this test with 5 actors per thread and it turns out this was quite important to increase probability of catching certain problems. I'd recommend to leave it this way for this particular test.

File: kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt (hunk 6 above)
Comment (ndkoval) [reply kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt]: Technically: yes, we can. For example, we can either always return `false` for other channels or create a special test for rendezvous one. However, it makes the code uglier. I would NOT test `isEmpty` since we have never had a test for it, and everything will be fixed with new channels.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt (hunk 10 above)
Comment (ndkoval) [reply kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: This code is used only for verification, to specify a state machine. Unfortunately, _lincheck_ does not have a mechanism to validate the data structure at the end of each invocation; we can add this feature to _licheck_.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt (hunk 11 above)
Comment (ndkoval) [reply kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: OK.

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt (hunk 7 above)
Comment (ndkoval) [reply kotlinx-coroutines-core/jvm/test/linearizability/LockFreeTaskQueueLCStressTest.kt]: We have tests with `removeFirstOrNull`

File: kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt (hunk 10 above)
Comment (elizarov) [reply kotlinx-coroutines-core/jvm/test/linearizability/LockFreeListLCStressTest.kt]: We don't have to wait for new lincheck. We can add it later.

**Comment by elizarov** [comment]
Still fails on tests.

**Comment by ndkoval** [comment]
I also added a sequential channel implementation
